File: haskell/triangle/src/Triangle.hs
module Triangle (TriangleType (..), triangleType) where

data TriangleType
  = Equilateral
  | Isosceles
  | Scalene
  | Illegal
  deriving (Eq, Show)

triangleType :: (Num a, Ord a) => a -> a -> a -> TriangleType
triangleType a b c
  | a + b < c || a + c < b || b + c < a = Illegal
  | a <= 0 || b <= 0 || c <= 0 = Illegal
  | a == b && b == c = Equilateral
  | a == b || b == c || a == c = Isosceles
  | a /= b && b /= c = Scalene
  | otherwise = Illegal

File: prolog/triangle/triangle.pl
triangular(Side1, Side2, Side3) :-
    Side1 > 0, Side2 > 0, Side3 > 0,
    Side1 + Side2 >= Side3, Side1 + Side3 >= Side2, Side3 + Side2 >= Side1.

triangle(Side1, Side2, Side3, "equilateral") :-
    Side1 = Side2, Side2 = Side3, triangular(Side1, Side2, Side3), !.

triangle(Side1, Side2, Side3, "isosceles") :-
    (Side1 = Side2; Side2 = Side3; Side1 = Side3),
    triangular(Side1, Side2, Side3), !.

triangle(Side1, Side2, Side3, "scalene") :-
    not(Side1 = Side2), not(Side2 = Side3), not(Side3 = Side1),
    triangular(Side1, Side2, Side3).
